### ISSUE TITLE
No matching certificate returned if servername contains a colon

### ISSUE THREAD

**Issue by LeCoyote** [libressl-portable/portable#660]
Bug found while using Unrealircd built against LibreSSL (3.2.5 and 3.3.2, with identical results).
`openssl s_client -connect localhost:6697` works fine, ie. a connection is made and the server returns a certificate as configured.
`openssl s_client -connect localhost:6697 -servername 2001:`
`140019933402432:error:14094417:SSL routines:ssl3_read_bytes:sslv3 alert illegal parameter:../ssl/record/rec_layer_s3.c:1543:SSL alert number 47`
`---`
`no peer certificate available`

The problem cannot be reproduced with OpenSSL 1.1.1
More detailed information at https://bugs.unrealircd.org/view.php?id=5863

**Comment by LeCoyote** [comment]
FWIW, an IDN as servername will also trigger the exact same reaction.

**Comment by botovq** [comment]
The -servername argument of the s_client sends an SNI extension which
should be an FQDN in ASCII encoding without trailing dot. If you want to
send an IDN, it must be in punycode. See RFC 6066 section 3 for details.

We do not do a full check for an A-label as that's too complex.
Whenever a LibreSSL server sees an SNI that doesn't conform to its
simple check for a valid hostname, it sends an illegal parameter alert,
which leads to the error you see.

While we do not explicitly check for that, literal IPv4 and IPv6
addresses are not allowed in an SNI extension. As you noted, IPv6
addresses won't make it through the server side check.

The openssl s_client tool (both in LibreSSL and OpenSSL) sends whatever
you specify in -servername in the SNI extension regardless of whether
that conforms to the spec. I have not carefully looked at hexchat or
weechat, but I would expect them to behave similarly.

Apparently, we're stricter than OpenSSL when checking the SNI server
side, but a priori this is not a bug.

**Comment by LeCoyote** [comment]
Thanks for the feedback. Let me rephrase from the user's point of view and see if I understood correctly.
IRC clients (the one I've used in this case) send an SNI that is not an FQDN in ASCII. Since OpenSSL is probably much more common on IRC servers, this deviation from the standards usually goes unnoticed. However, LibreSSL does what it is supposed to do, hence the IRC clients in question should be the ones to address the issue since, as far as I can tell, the user has no choice as to what is send as the SNI.
Is that correct?

**Comment by botovq** [comment]
Yes, this is correct: If the chat client is configured to talk to a
server at a certain IP address, the client should not send that address
in their SNI.

The clients should either not send an SNI at all (which may break with
TLSv1.3 servers that choose to enforce presence of an SNI - I don't know
how common that is among chat servers) or they should allow the user to
specify a valid hostname to send in that extension.

**Comment by LeCoyote** [comment]
Not sending an SNI seems to be fine with the server in question, and sounds like a sane solution anyway.

Many thanks! Closing with this comment as it is, indeed, not a bug. Cheers!
